CONTRIBUTING.md

```diff
@@ -106,3 +106,67 @@ To update it, run:
 npm install puppeteer
 node scripts/build_reserved.js
 ```
+
+## A whirlwind tour to the compiler codebase
+
+### Folder structure:
+
+- `bin` contains all binaries shipped in a Melange distribution.
+- `docs` contains the old [BuckleScript
+  manual](https://melange.re/melange/Manual.html) which we host to consult
+  (rarely).
+- `jscomp` is where the compiler implementation and some tests live:
+    - `common` defines the Melange `common` private library, housing code
+      shared by both the Melange binary and the `melange.ppx` library.
+    - `core` defines the `core` Melange library, containing the bulk of the
+      compiler backend implementation.
+    - `ext` contains the sources for the `ext` private library. This is a
+      standard library extension with additional functions and data structutres
+      used throughout the Melange code.
+    - `js_parser` is a vendored copy of the [Flow
+      parser](https://github.com/facebook/flow/tree/main/src/parser), used in
+      Melange to classify `%mel.raw` JavaScript code.
+    - `runtime` is the Melange runtime. It has:
+        - the `melange.js` library.
+        - the `Caml_*` modules, implementing the low-level OCaml primitives.
+          While these are technically part of the `melange.js` library, they:
+            1. aren't exposed in the `Js.*` module.
+            2. are only ever accessed from JS: Melange internals don't address
+               any of these modules directly, and you shouldn't either.
+            3. are still exposed e.g. as `Js__Caml_array` because of Dune
+               namespacing implementation details.
+    - `stdlib` is the Melange stdlib, included in the final distribution.
+      - Depends on the runtime
+    - `others` contains the sources for the additional opt-in libraries
+      distributed with Melange: `melange.belt`, `melange.dom` and
+      `melange.node`.
+    - `test` has both:
+        - the sources for the Melange runtime tests that are executed on every
+          CI run
+        - a snapshot of their compilation to JavaScript (in
+          [`jscomp/test/dist`](https://github.com/melange-re/melange/tree/main/jscomp/test/dist)
+          and
+          [`jscomp/test/dist-es6`](https://github.com/melange-re/melange/tree/main/jscomp/test/dist-es6))
+            - **NOTE**: these snapshots are currently built manually. To do so,
+              comment the only line in
+              [`jscomp/dune`](https://github.com/melange-re/melange/blob/main/jscomp/dune)
```

Comment:
Maybe worth pointing to the current commit in `main` now, so that at least if folder layout changes in the future, the docs can point to the state where things were written (applies to all links to code in the document).

Reply:
I was thinking the fact they're on `main` will let us catch if we ever break them. Meaning the guide should be updated.

Reply:
> I was thinking the fact they're on `main` will let us catch if we ever break them.

I doubt we will know we are breaking them, as there is no mechanism to warn us.